### PR TITLE
:neckbeard: QNN toolkit

### PR DESCRIPTION
**Pull request by trajepl**
## Describe your changes

QNN toolkit

CI e2e test for qnn mobile net: https://github.com/microsoft/Olive/pull/874/checks?check_run_id=20488489220

## Checklist before requesting a review
- [ ] Add unit tests for this change.
- [ ] Make sure all tests can pass.
- [ ] Update documents if necessary.
- [ ] Lint and apply fixes to your code by running `lintrunner -a`
- [ ] Is this a user-facing change? If yes, give a description of this change to be included in the release notes.

### CODE REVIEW

File: olive/evaluator/olive_evaluator.py [hunk 1]
```diff
@@ -1040,6 +1041,75 @@ def _evaluate_raw_latency(
...
+        self,
+        model: QNNModelHandler,
+        metric: Metric,
+        dataloader: Dataset,
+        post_func=None,
+        device: Device = Device.CPU,
+        execution_providers: Union[str, List[str]] = None,
+    ) -> Tuple[OliveModelOutput, Any]:
+        dataloader = self._prepare_dataloader(dataloader, model)
+        session = model.prepare_session(inference_settings=self.get_inference_settings(metric), device=device)
```
Comment (guotuofeng) [inline olive/evaluator/olive_evaluator.py:1058]: metric.get_inference_settings

**Comment by guotuofeng** [comment]
not checkin the binray files such .raw

File: olive/model/handler/qnn.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+                elif platform.system() == "Windows":
+                    # might be different for arm devices
+                    lib_targets = "x64"
+                    model_lib_suffix = ".dll"
+            model_folder = Path(model_path) / lib_targets
+            model_paths = list(model_folder.glob(f"*{model_lib_suffix}"))
+            if not model_paths:
+                raise FileNotFoundError(f"No model file found in {model_folder}")
+            elif len(model_paths) > 1:
+                raise FileNotFoundError(f"Multiple model files found in {model_folder}: {model_paths}")
```
Comment (guotuofeng) [inline olive/model/handler/qnn.py:74]: I guess the exception type is confused.

File: olive/passes/qnn/context_binary_generator.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+
+    def _run_for_config(
+        self,
+        model: Union[TensorFlowModelHandler, PyTorchModelHandler, ONNXModelHandler],
+        data_root: str,
+        config: Dict[str, Any],
+        output_model_path: str,
+    ) -> QNNModelHandler:
+        main_cmd = "qnn-context-binary-generator"
+        runner = QNNSDKRunner(dev=True)
```
Comment (guotuofeng) [inline olive/passes/qnn/context_binary_generator.py:60]: what's meaning dev=True for all cases?

Comment (trajepl) [reply olive/passes/qnn/context_binary_generator.py:60]: `dev=True` means run the cmd locally. Run QNN optimization has following steps:
1. convert and build model in x64 windows/linux (dev = True
2. set the model to target device(arm-windows or android) (dev=False

File: olive/model/handler/qnn.py [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+                elif platform.system() == "Windows":
+                    # might be different for arm devices
+                    lib_targets = "x64"
+                    model_lib_suffix = ".dll"
+            model_folder = Path(model_path) / lib_targets
+            model_paths = list(model_folder.glob(f"*{model_lib_suffix}"))
+            if not model_paths:
+                raise FileNotFoundError(f"No model file found in {model_folder}")
+            elif len(model_paths) > 1:
+                raise RuntimeError(f"Multiple model files found in {model_folder}: {model_paths}")
```
Comment (guotuofeng) [inline olive/model/handler/qnn.py:74]: should the optimization pass since the RuntimeError will be ignored? or need we raise assertionerror so the optimization will fail?

File: olive/passes/qnn/context_binary_generator.py (hunk 3 above)
Comment (guotuofeng) [reply olive/passes/qnn/context_binary_generator.py:60]: If it is used for run locally or remotely, suggest to change the name to local_run or something else. 

In my mind, dev=True means run the optimization using the local branch Olive bits.

Comment (trajepl) [reply olive/passes/qnn/context_binary_generator.py]: Updated with `optional_local_run` where the True value is used to tell current runner can run cmd locally.
The value is set to False only if the `cmd` have to be run remotely but cannot in local env.

File: olive/model/handler/qnn.py (hunk 4 above)
Comment (trajepl) [reply olive/model/handler/qnn.py]: Make sense. Updated both case with assertion.
image

File: olive/platform_sdk/qualcomm/env.py [hunk 5]
```diff
@@ -11,7 +11,7 @@
 
 
 class SDKEnv:
-    def __init__(self, sdk: str, root_env_name: str, target_arch: str = None, dev: bool = False):
+    def __init__(self, sdk: str, root_env_name: str, target_arch: str = None, optional_local_run: bool = False):
```
Comment (jambayk) [inline olive/platform_sdk/qualcomm/env.py:14]: I think a name like `use_dev_tools` is a better name than `optional_local_run`. Even an inference can be run locally.

File: olive/evaluator/olive_evaluator.py [hunk 6]
```diff
@@ -1040,6 +1041,79 @@ def _evaluate_raw_latency(
...
+
+        preds = []
+        targets = []
+        logits = []
+        for data_dir, input_list, labels in dataloader:
+            result = session(input_list, data_dir).get("result")
+            if post_func:
+                outputs = post_func(result)
+            else:
+                raise ValueError("Post processing function is required for QNN model")
```
[the remark this answers is not in the document]
Comment (trajepl) [reply olive/evaluator/olive_evaluator.py:1070]: Not a hard requirement. I am planning to update this part in following PR.

File: olive/passes/qnn/model_lib_generator.py [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+from olive.platform_sdk.qualcomm.runner import QNNSDKRunner
+
+logger = logging.getLogger(__name__)
+
+
+class QNNModelLibGenerator(Pass):
+    @staticmethod
+    def _default_config(accelerator_spec: AcceleratorSpec) -> Dict[str, PassConfigParam]:
+        return {
+            "use_olive_env": PassConfigParam(
```
[the remark this answers is not in the document]
Comment (trajepl) [reply olive/passes/qnn/model_lib_generator.py]: Updated.

File: olive/platform_sdk/qualcomm/env.py (hunk 5 above)
Comment (trajepl) [reply olive/platform_sdk/qualcomm/env.py]: Updated.

File: examples/test/test_qnn_tooklit.py
[the remark this answers is not in the document]
Comment (trajepl) [reply examples/test/test_qnn_tooklit.py:1]: Fixed with rearrange the input data shape.
Will try RawDataLoader to automatic the shape transposing in following changes.
Thanks!

**Review by jambayk — APPROVED** [review]
